### PR TITLE
Ver 5.0.0 Feedback Thread

### PR DESCRIPTION
**Pull request by black7375**
**Describe the PR**

Build using SCSS, and have optional systems.

---

Setup - Please refer [Wiki:Installation Guide](https://github.com/black7375/Firefox-UI-Fix/wiki/Installation-Guide)

1. Download files
```shell
## First user ##################################################################
# At <FIREFOX_PROFILE_DIRECTORY>

# Original dev version
git clone https://github.com/black7375/Firefox-UI-Fix -b dev

# Photon style dev version
git clone https://github.com/black7375/Firefox-UI-Fix  chrome -b photon-style-dev

# Proton style dev version
git clone https://github.com/black7375/Firefox-UI-Fix chrome -b proton-style-dev

## Existing user ###############################################################
# At <FIREFOX_PROFILE_DIRECTORY>/chrome

# Original dev version
git checkout dev
git pull

# Proton style dev version
git checkout photon-style-dev
git pull

# Proton style dev version
git checkout proton-style-dev
git pull
```
2. Copy `user.js` to profile directory
    :warning: You must copy `user.js` file before the theme works.
    The option system depends on user configuration, and ***nothing applies without settings***.
3. Restart firefox

---

Customize

You can be customized using the following file:
- https://github.com/black7375/Firefox-UI-Fix/blob/dev/user.js

:warning: However, there are several caveats.
1. File `user.js` must modify the file located in `<FIREFOX_PROFILE>/user.js`, not `<FIREFOX_PROFILE>/chrome/user.js`.
2. It can be configurable with `user.js` file and `about:config` and needs to be restarted.
  The option system is built using [`CSS -moz-bool-pref() @supports function`](https://udn.realityripple.com/docs/Mozilla/Gecko/Chrome/CSS/-moz-bool-pref) as a base.
  `@supports` change in CSS is not detected in real time. (Only start time)
  So a restart is required, and if the mozilla need real time changes, are using `@media` to handle it.
  However, since this project only use pure CSS, we cannot add `@media rules`.
  - [Bug 1267890 - Support detecting bool preferences in chrome stylesheets](https://bugzilla.mozilla.org/show_bug.cgi?id=1267890)
  - [Bug 1698132 -  Improve caching behaviour of -moz-bool-pref](https://bugzilla.mozilla.org/show_bug.cgi?id=1698132)
3. After editing `user.js` file, click the `Clear startup cache...` in `about:support`.
  If the `Clear startup cache...` button is not used, the setting may be overridden.
4. Even if you change the setting in `about:config`, it will not take effect if it remains in `user.js`.
  It is recommended to explicitly change the value in the `user.js` file.
5. If you remove a setting from the `user.js` file, the last value is preserved.
  It is recommended to explicitly change the value to `false` in the `user.js` file.

---

**Related Issue**

Main

- SCSS #234 
- Option #267 #347
- Documents https://github.com/black7375/Firefox-UI-Fix/commit/cbf14cd55a9edada7ab2f5f1b626608fb9fe38a2

Additional

- [x]  Megabar css tweaks option. #348
- [x]  Not all hamburger menu icons show in photon-tab style #329 
- [x] Active pinned tabs are wider than inactive ones #38 

Frequency user requirements https://github.com/black7375/Firefox-UI-Fix/pull/351/commits/cc75c6a75424e509a3fec7d2d6069980f99a2370 https://github.com/black7375/Firefox-UI-Fix/pull/351/commits/1943c5cfc6fc2d81e19c21e4cbf3b440e565acb1

- [x]  Photon style panel is missing icons present in the old design #308
- [x] Color
  - [x]  When photon style is used, change light blue to darker blue #202 
  - [x]  Accent color on Dev edition? #76 
- [x]  Rounded corners (Windows) #266 
- [x]  Photon-style changes to emulate old design #222 

Customize
- [x] Move lepton stylesheets into their own files #352
- [x] `userChrome-override.css`, `userContent-override.css` & installer support
- [x] `user-override.js` & installer support

**Screenshots**

image

Using `dev` branch and the following settings

- `userChrome.tab.on_bottom`
- `userChrome.tab.on_bottom.menubar_on_top`

**Additional context**

You can suggest compatibility issues, option names, etc.

### CODE REVIEW

File: user.js [hunk 1]
```diff
@@ -26,6 +26,134 @@ user_pref("browser.newtabpage.activity-stream.improvesearch.handoffToAwesomebar"
 // Browser Theme Based Scheme - Will be activate 95 Above
 // user_pref("layout.css.prefers-color-scheme.content-override", 3);
 
+// ** Theme Related Options ****************************************************
+// == Theme Distribution Settings ==============================================
+user_pref("userChrome.tab.connect_to_window",          true); // Original, Photon
```
Comment (hdk5) [inline user.js:31]: Would it be possible to have option names that are more distinguishable from builtins? (e.g. `black7375.lepton.userChrome.tab.connect_to_window`)

Comment (black7375) [reply user.js:31]: I use `userChrome.` and `userContent` to keep it simple while distinguishing it from the built-in API.
If you search for it in `about:config` settings, it is an unused name.

**Comment by roland-rollo** [comment]
I created `chrome` folder with
```
css
icons
src
userChrome.css
userContent.css
```

And with `user.js` in main profile folder and everything works without problem. Looks great! Every change I made to `user.js` worked and were clear how to use.

Does this files are needed in `chrome` folder?
```
docs
package.json
yarn.lock
```

**Comment by black7375** [comment]
They contain the information needed for development.

This document may be helpful.
https://github.com/black7375/Firefox-UI-Fix/blob/dev/docs/Project_Structure.md

**Comment by MonfGeiger** [comment]
so let me try to parse this one more time: we could try and change the userchrome flags through about:config all we want, but no matter what, those values will NEVER truly (at this moment) apply that way. we would have to edit those flags into the user.js file (if not already included) and set them that way after following the steps in this feedback thread.

that is.....frankly annoying. I thought we could set the values through firefox itself by now, but we still have to do it manually, or edit the userchrome folder directly. my apologies for my griping, but will we ever truly be able to set userchrome flags through firefox's about:config page any time soon, or is that just never happening no matter what?

**Comment by black7375** [comment]
@MonfGeiger 

It's possible using only `about:config`.
The limit of using `about:config` is that it is initialized every time with the settings in the `user.js` file.

What we can see here:
1. If the setting is not enabled by default, you can only use `about:config`.
2. If you need to touch the default settings, you can delete `user.js` after running once.

This is why the wiki document is written to delete the file.
https://github.com/black7375/Firefox-UI-Fix/wiki/Options#using-aboutconfigs

**Comment by MonfGeiger** [comment]
.....my potato brain did not see that until now. forgive my big brain move, I have indeed deleted the thing and I see it working again

**Pull request merged**